File: apps/aiyou_stack/aiyou-fastapi-services/tests/integration/test_refactorer.mjs
#!/usr/bin/env node
/**
 * Integration tests for Code Refactorer Agent
 *
 * Tests the refactorer's analyze and refactor capabilities
 */

import { refactorCode, analyzeCode } from "../../tools/orchestrator/ace_with_refactor.mjs";
import assert from "node:assert";

// Test fixtures
const sampleCode = {
  javascript: `
function calc(a,b,c) {
  var result = 0;
  if(a>0) {
    if(b>0) {
      result = a+b+c
    }
  }
  return result
}
`,
  python: `
def calc(a,b,c):
    result = 0
    if a>0:
        if b>0:
            result = a+b+c
    return result
`,
};

// Test suite
const tests = {
  async testAnalyzeJavaScript() {
    console.log("Testing: analyzeCode() with JavaScript...");
    const analysis = await analyzeCode(sampleCode.javascript, "javascript");

    assert(analysis.issues, "Analysis should return issues array");
    assert(analysis.metrics, "Analysis should return metrics");
    assert(analysis.recommendations, "Analysis should return recommendations");
    assert(typeof analysis.metrics.complexity === "number", "Complexity should be a number");
    assert(
      typeof analysis.metrics.maintainability === "number",
      "Maintainability should be a number",
    );

    console.log("✓ analyzeCode() returns valid structure");
    return true;
  },

  async testRefactorJavaScript() {
    console.log("Testing: refactorCode() with JavaScript...");
    const result = await refactorCode(sampleCode.javascript, {
      language: "javascript",
      focus: ["readability", "best-practices"],
      aggressiveness: "moderate",
    });

    assert(result.refactoredCode, "Refactoring should return refactored code");
    assert(result.summary, "Refactoring should return summary");
    assert(Array.isArray(result.improvements), "Improvements should be an array");
    assert(result.improvements.length > 0, "Should have at least one improvement");

    console.log("✓ refactorCode() returns valid refactoring result");
    return true;
  },

  async testRefactorWithFocusAreas() {
    console.log("Testing: refactorCode() with different focus areas...");

    const focusAreas = [
      ["readability"],
      ["performance"],
      ["best-practices"],
      ["readability", "performance"],
    ];

    for (const focus of focusAreas) {
      const result = await refactorCode(sampleCode.javascript, {
        language: "javascript",
        focus,
        aggressiveness: "conservative",
      });

      assert(result.refactoredCode, `Should refactor with focus: ${focus.join(", ")}`);
    }

    console.log("✓ refactorCode() works with various focus areas");
    return true;
  },

  async testRefactorAggressiveness() {
    console.log("Testing: refactorCode() with different aggressiveness levels...");

    const levels = ["conservative", "moderate", "aggressive"];

    for (const level of levels) {
      const result = await refactorCode(sampleCode.javascript, {
        language: "javascript",
        aggressiveness: level,
      });

      assert(result.refactoredCode, `Should refactor with aggressiveness: ${level}`);
    }

    console.log("✓ refactorCode() supports all aggressiveness levels");
    return true;
  },

  async testAnalyzePython() {
    console.log("Testing: analyzeCode() with Python...");
    const analysis = await analyzeCode(sampleCode.python, "python");

    assert(analysis.issues, "Python analysis should return issues");
    assert(analysis.metrics, "Python analysis should return metrics");

    console.log("✓ analyzeCode() works with Python code");
    return true;
  },

  async testRefactorErrorHandling() {
    console.log("Testing: error handling with invalid code...");

    try {
      // Test with empty code
      const result = await refactorCode("", {
        language: "javascript",
      });

      // Should still return a valid structure even with empty code
      assert(result.refactoredCode !== undefined, "Should handle empty code gracefully");
      console.log("✓ Handles empty code gracefully");
    } catch (error) {
      console.log("✓ Properly throws error for invalid input");
    }

    return true;
  },

  async testConfigurationDefaults() {
    console.log("Testing: default configuration values...");

    // Test without explicit config
    const result = await refactorCode(sampleCode.javascript);

    assert(result.refactoredCode, "Should use default configuration");
    console.log("✓ Uses sensible defaults when config not provided");
    return true;
  },
};

// Test runner
async function runTests() {
  console.log("=".repeat(60));
  console.log("Code Refactorer Integration Tests");
  console.log("=".repeat(60));
  console.log();

  let passed = 0;
  let failed = 0;
  const failedTests = [];

  for (const [testName, testFn] of Object.entries(tests)) {
    try {
      await testFn();
      passed++;
    } catch (error) {
      failed++;
      failedTests.push({ name: testName, error: error.message });
      console.error(`✗ ${testName} failed:`, error.message);
    }
    console.log();
  }

  console.log("=".repeat(60));
  console.log("Test Results");
  console.log("=".repeat(60));
  console.log(`Total: ${passed + failed}`);
  console.log(`Passed: ${passed}`);
  console.log(`Failed: ${failed}`);

  if (failed > 0) {
    console.log("\nFailed tests:");
    failedTests.forEach(({ name, error }) => {
      console.log(`  - ${name}: ${error}`);
    });
    process.exit(1);
  }

  console.log("\n✓ All tests passed!");
  console.log("=".repeat(60));
}

// Run tests if executed directly
if (import.meta.url === `file://${process.argv[1]}`) {
  runTests().catch((error) => {
    console.error("Test runner error:", error);
    process.exit(1);
  });
}

export { tests, runTests };

File: apps/aiyou_stack/aiyou-fastapi-services/tests/integration/test_refactorer.py
#!/usr/bin/env python3
"""Integration tests for Code Refactorer Agent (Python version)

Tests the refactorer's analyze and refactor capabilities from Python.
"""

import asyncio
import os
import sys

# Add parent directory to path for imports
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "../..")))

# Sample test fixtures
SAMPLE_CODE = {
    "javascript": """
function calc(a,b,c) {
  var result = 0;
  if(a>0) {
    if(b>0) {
      result = a+b+c
    }
  }
  return result
}
""",
    "python": """
def calc(a,b,c):
    result = 0
    if a>0:
        if b>0:
            result = a+b+c
    return result
""",
}


class TestRefactorer:
    """Integration test suite for Code Refactorer"""

    def __init__(self):
        self.passed = 0
        self.failed = 0
        self.failed_tests = []

    async def test_analyze_code(self):
        """Test code analysis functionality"""
        print("Testing: Code analysis...")

        # Mock analysis for now (until Python refactorer is fully implemented)
        # In real implementation, this would call the actual refactorer
        analysis = {
            "issues": [
                {
                    "severity": "medium",
                    "type": "complexity",
                    "description": "Nested conditionals",
                    "suggestion": "Flatten logic",
                },
            ],
            "metrics": {"complexity": 45, "maintainability": 75, "technical_debt": "medium"},
            "recommendations": ["Add type hints", "Improve naming"],
        }

        assert "issues" in analysis, "Analysis should return issues"
        assert "metrics" in analysis, "Analysis should return metrics"
        assert isinstance(analysis["metrics"]["complexity"], int), "Complexity should be int"

        print("✓ Code analysis returns valid structure")
        return True

    async def test_refactor_code(self):
        """Test code refactoring functionality"""
        print("Testing: Code refactoring...")

        # Mock refactoring result
        result = {
            "refactored_code": "# Refactored code here",
            "summary": "Improvements made",
            "improvements": ["Better naming", "Type hints added", "Reduced complexity"],
        }

        assert "refactored_code" in result, "Should return refactored code"
        assert "summary" in result, "Should return summary"
        assert isinstance(result["improvements"], list), "Improvements should be list"
        assert len(result["improvements"]) > 0, "Should have improvements"

        print("✓ Code refactoring returns valid result")
        return True

    async def test_focus_areas(self):
        """Test refactoring with different focus areas"""
        print("Testing: Different focus areas...")

        focus_areas = [
            ["readability"],
            ["performance"],
            ["best-practices"],
            ["readability", "performance"],
        ]

        for focus in focus_areas:
            # Mock refactoring with focus
            result = {"refactored_code": f"# Refactored with focus: {', '.join(focus)}"}
            assert "refactored_code" in result, f"Should work with focus: {focus}"

        print("✓ Refactoring works with various focus areas")
        return True

    async def test_aggressiveness_levels(self):
        """Test refactoring with different aggressiveness levels"""
        print("Testing: Aggressiveness levels...")

        levels = ["conservative", "moderate", "aggressive"]

        for level in levels:
            # Mock refactoring with aggressiveness level
            result = {"refactored_code": f"# Refactored with level: {level}"}
            assert "refactored_code" in result, f"Should work with level: {level}"

        print("✓ Supports all aggressiveness levels")
        return True

    async def test_error_handling(self):
        """Test error handling with invalid inputs"""
        print("Testing: Error handling...")

        # Test with empty code
        try:
            result = {"refactored_code": ""}  # Mock empty result
            assert result["refactored_code"] is not None, "Should handle empty code"
            print("✓ Handles edge cases gracefully")
        except Exception as e:
            print(f"✓ Properly handles errors: {e}")

        return True

    async def run_test(self, test_name, test_fn):
        """Run a single test and track results"""
        try:
            await test_fn()
            self.passed += 1
        except AssertionError as e:
            self.failed += 1
            self.failed_tests.append({"name": test_name, "error": str(e)})
            print(f"✗ {test_name} failed: {e}")
        except Exception as e:
            self.failed += 1
            self.failed_tests.append({"name": test_name, "error": str(e)})
            print(f"✗ {test_name} error: {e}")
        print()

    async def run_all_tests(self):
        """Run all integration tests"""
        print("=" * 60)
        print("Code Refactorer Integration Tests (Python)")
        print("=" * 60)
        print()

        # Get all test methods
        test_methods = [
            (name, getattr(self, name))
            for name in dir(self)
            if name.startswith("test_") and callable(getattr(self, name))
        ]

        # Run each test
        for test_name, test_fn in test_methods:
            await self.run_test(test_name, test_fn)

        # Print results
        print("=" * 60)
        print("Test Results")
        print("=" * 60)
        print(f"Total: {self.passed + self.failed}")
        print(f"Passed: {self.passed}")
        print(f"Failed: {self.failed}")

        if self.failed > 0:
            print("\nFailed tests:")
            for test in self.failed_tests:
                print(f"  - {test['name']}: {test['error']}")
            return False

        print("\n✓ All tests passed!")
        print("=" * 60)
        return True


async def main():
    """Main test runner"""
    suite = TestRefactorer()
    success = await suite.run_all_tests()
    raise SystemExit(0 if success else 1)


if __name__ == "__main__":
    asyncio.run(main())
